backend/pkg/database/refresh_materialized_views.go

```diff
@@ -0,0 +1,129 @@
+	"gorm.io/gorm"
+)
+
+const DefaultViewRefreshIntervalSecond = 5
+
+type refreshFunc func(db *gorm.DB, viewNames []string)
+
+type ViewRefreshHandler struct {
+	mu                        sync.Mutex
+	viewsToRefresh            map[string][]string // map of tables that shows which views should be refreshed
+	tableChanged              map[string]bool
+	refreshFunc               refreshFunc
+	viewRefreshIntervalSecond time.Duration
+}
+
+func (vh *ViewRefreshHandler) TableChanged(table string) {
+	vh.mu.Lock()
+	defer vh.mu.Unlock()
+	vh.tableChanged[table] = true
+}
+
+func (vh *ViewRefreshHandler) GetAndClearChanges() map[string]bool {
+	vh.mu.Lock()
+	defer vh.mu.Unlock()
+	tables := vh.tableChanged
+	vh.tableChanged = make(map[string]bool)
+	return tables
+}
+
+func (vh *ViewRefreshHandler) runRequiredRefreshes(db *gorm.DB) {
+	viewNames := vh.getViewsToRefresh()
+	vh.refreshFunc(db, viewNames)
+}
+
+func (vh *ViewRefreshHandler) getViewsToRefresh() []string {
+	viewToRefresh := make(map[string]bool)
+	tables := vh.GetAndClearChanges()
+	for table, ok := range tables {
+		if ok {
+			for _, viewName := range vh.viewsToRefresh[table] {
+				viewToRefresh[viewName] = true
+			}
+		}
+	}
+	views := make([]string, 0)
+	for viewName, ok := range viewToRefresh {
+		if ok {
+			views = append(views, viewName)
+		}
+	}
+	return views
+}
+
+func (vh *ViewRefreshHandler) RegisterViewRefreshHandler(f refreshFunc) {
+	vh.refreshFunc = f
+}
+
+func (vh *ViewRefreshHandler) IsSetViewRefreshHandler() bool {
+	return vh.refreshFunc != nil
+}
+
+func createViewsToRefreshByTable() map[string][]string {
+	viewsToRefresh := make(map[string][]string)
+	tables := []string{
+		applicationTableName,
+		resourceTableName,
+		packageTableName,
+		vulnerabilityTableName,
+	}
+	for _, table := range tables {
+		viewsToRefresh[table] = materializedViews
+	}
+
+	return viewsToRefresh
+}
+
+func (db *Handler) SetMaterializedViewHandler(config *DBConfig) {
+	db.ViewRefreshHandler = &ViewRefreshHandler{
+		tableChanged:              make(map[string]bool),
+		viewsToRefresh:            createViewsToRefreshByTable(),
+		viewRefreshIntervalSecond: time.Duration(config.ViewRefreshIntervalSecond) * time.Second,
+	}
+}
+
+func (db *Handler) RefreshMaterializedViews() {
+	for {
+		<-time.After(db.ViewRefreshHandler.viewRefreshIntervalSecond)
+		db.ViewRefreshHandler.runRequiredRefreshes(db.DB)
+	}
+}
+
+func RefreshMaterializedViews(db *gorm.DB, viewNames []string) {
+	for _, viewName := range viewNames {
+		if err := db.Exec(fmt.Sprintf(refreshMaterializedViewConcurrentlyCommand, viewName)).Error; err != nil {
+			log.Errorf("Failed to refresh materialized %s: %v", viewName, err)
+		}
+	}
+}
+
+func initMaterializedViews(db *gorm.DB, viewNames []string) {
+	for _, viewName := range viewNames {
+		if err := db.Exec(fmt.Sprintf(refreshMaterializedViewCommand, viewName)).Error; err != nil {
+			log.Fatalf("Failed to init materialized %s: %v", viewName, err)
+		}
+	}
+}
```

Comment:
We could remove this function and instead have the createAllViews function in the DB just mark the tables as changed, i.e.
```
func createAllViews(...) {
        ....
         
	if err := db.Exec(fmt.Sprintf(applicationsViewQuery, createCommand)).Error; err != nil {
		log.Fatalf("Failed to create applications_view: %v", err)
	}
	if dbDriver == DBDriverTypePostgres {
	       // Register postgres view refresh *if* applications, resources or packages tables change
	       // those are the only tables that view requires.
	       dbHandler.ViewRefreshHandler.RegisterViewRefreshHandler(
	               func(db *gorm.DB){
	                       if err := db.Exec(fmt.Sprintf(refreshMaterializedViewConcurrentlyCommand, applicationsViewName)).Error; err != nil {
	                               log.Errorf("Failed to refresh materialized %s: %v", applicationsViewName, err)
		               }
                       },
                       "applications",
                       "resources",
                       "packages",
               )
               // Mark the tables the view is interested in as changed so that the view refresh
               // handler will refresh this view at least once on the next cycle even though the
               // tables have not really changed
               dbHandler.ViewRefreshHandler.TableChanged("applications")
               dbHandler.ViewRefreshHandler.TableChanged("resources")
               dbHandler.ViewRefreshHandler.TableChanged("packages")
	}
	
	....
}
```

Reply:
In the case of all related tables not being populated, refreshing the view concurrently will fail. So this is the reason for the first time we need to refresh it without the `CONCURRENTLY` option.